<a href="https://colab.research.google.com/github/aashish-1904/sampling-assignment/blob/main/sampling_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Sampling Assignment

## Loading Datasets and Libraries

In [35]:
!git clone https://github.com/AnjulaMehto/Sampling_DataSet.git

fatal: destination path 'Sampling_DataSet' already exists and is not an empty directory.


In [36]:
import pandas as pd
import numpy as np
from sklearn import svm
from collections import Counter
from imblearn.over_sampling import SMOTE
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,f1_score,precision_score, recall_score

In [37]:
df = pd.read_csv("Sampling_DataSet/Creditcard_data.csv")

## Oversampling using SMOTE

In [38]:
df['Class'].value_counts()

0    763
1      9
Name: Class, dtype: int64

In [39]:
from sklearn.model_selection import train_test_split
X = df.loc[ : , df.columns != 'Class']
y = df.Class
# split into 70:30 ration
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0)
  
# describes info about train and test set
print("Number transactions X_train dataset: ", X_train.shape)
print("Number transactions y_train dataset: ", y_train.shape)
print("Number transactions X_test dataset: ", X_test.shape)
print("Number transactions y_test dataset: ", y_test.shape)

Number transactions X_train dataset:  (540, 30)
Number transactions y_train dataset:  (540,)
Number transactions X_test dataset:  (232, 30)
Number transactions y_test dataset:  (232,)


In [40]:
print('before SMOTE:',Counter(y_train))
from imblearn.over_sampling import SMOTE
sm = SMOTE(sampling_strategy='minority')
X_train2, Y_train2 = sm.fit_resample(X_train, y_train)
print('After SMOTE:',Counter(Y_train2))

before SMOTE: Counter({0: 533, 1: 7})
After SMOTE: Counter({0: 533, 1: 533})


In [41]:
X_train2.shape

(1066, 30)

In [42]:
Y_train2.shape

(1066,)

In [43]:
X_test.shape

(232, 30)

In [44]:
y_test.shape

(232,)

In [45]:
Class = Y_train2.append(y_test)

In [46]:
len(Class)

1298

In [47]:
df2 = pd.concat([X_train2, X_test])
df2.shape

(1298, 30)

In [48]:
df2['Class'] = Class

In [49]:
df2.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   247 -0.685770  0.554019  2.014829 -0.561033  0.683347  0.002232  0.804394   
1   427 -0.847312  0.854261  0.338816  0.890137  0.804751  1.165501 -0.081408   
2    26 -0.535388  0.865268  1.351076  0.147575  0.433680  0.086983  0.693039   
3   580  1.267030 -0.071114  0.037680  0.512683  0.242392  0.705212 -0.226582   
4   579  1.214170  0.210481  0.484651  0.479768 -0.261955 -0.527039  0.021782   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0 -0.154563  0.339900  ...  0.113873  0.642716 -0.421980 -0.411622  0.359461   
1  0.879014 -0.394737  ... -0.046690 -0.075301 -0.308479 -1.733137  0.087036   
2  0.179742 -0.285642  ...  0.049526  0.206537 -0.187108  0.000753  0.098117   
3  0.109483  0.657565  ... -0.164468 -0.177225 -0.222918 -1.245505  0.678360   
4 -0.106888 -0.037631  ... -0.224292 -0.594609  0.159877  0.091873  0.140964   

        V26       V27       V28  Amount  Class  
0 -0.545006 -0.040244 -0.096934   18.94      0  
1 -0.129209  0.294334  0.071198   11.36      0  
2 -0.553471 -0.078306  0.025427    1.77      0  
3  0.525059  0.002920 -0.003333   12.36      0  
4  0.227406 -0.017389  0.016030    5.98      0  

[5 rows x 31 columns]

In [50]:
df2.shape

(1298, 31)

In [51]:
df2['Class'].value_counts()

0    763
1    535
Name: Class, dtype: int64

## Models

In [52]:
from sklearn import metrics
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
def ml_model(X_train,X_test, Y_train, Y_test):
  MLA = [LogisticRegression(),KNeighborsClassifier(),DecisionTreeClassifier(),GaussianNB(),RandomForestClassifier(),svm.SVC()]
  MLA_columns = []
  MLA_compare = pd.DataFrame(columns = MLA_columns)
  row_index = 0
  for alg in MLA:
    predicted = alg.fit(X_train, Y_train).predict(X_test)
    MLA_name = alg.__class__.__name__
    MLA_compare.loc[row_index,'Model Name'] = MLA_name
    MLA_compare.loc[row_index, 'Train Accuracy'] = round(alg.score(X_train, Y_train), 2)
    MLA_compare.loc[row_index, 'Test Accuracy'] = round(alg.score(X_test, Y_test), 2)
    MLA_compare.loc[row_index, 'Precision'] = round(precision_score(Y_test, predicted),2)
    MLA_compare.loc[row_index, 'Recall'] = round(recall_score(Y_test, predicted),2)
    MLA_compare.loc[row_index, 'F1 score'] = round(f1_score(Y_test, predicted),2)
    row_index+=1
  MLA_compare.sort_values(by = ['Test Accuracy'], ascending = False, inplace = True)    
  return MLA_compare  

## Random Sampling

In [53]:
Z=1.96
p=0.4
E=0.05
n = (Z*Z*p*(1-p))/(E*E)
n

368.7935999999999

In [54]:
random_sample = df2.sample(n=369, random_state=0)

In [55]:
X = random_sample.loc[ : , random_sample.columns != 'Class']
Y = random_sample.Class

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=1)

ml_model(X_train, X_test, Y_train, Y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Model Name  Train Accuracy  Test Accuracy  Precision  Recall  \
4  RandomForestClassifier            1.00           1.00       1.00    1.00   
2  DecisionTreeClassifier            1.00           0.95       0.86    1.00   
0      LogisticRegression            0.97           0.94       0.88    0.95   
3              GaussianNB            0.93           0.92       0.81    1.00   
1    KNeighborsClassifier            0.85           0.77       0.62    0.89   
5                     SVC            0.71           0.71       0.57    0.66   

   F1 score  
4      1.00  
2      0.93  
0      0.91  
3      0.89  
1      0.73  
5      0.61

## Systematic Sampling

In [56]:
def systematic_sampling(df, step):
 
    indexes = np.arange(0, len(df), step=step)
    systematic_sample = df.iloc[indexes]
    return systematic_sample

In [57]:
systematic_sample = systematic_sampling(df2, 2)

In [58]:
X = systematic_sample.loc[ : , systematic_sample.columns != 'Class']
Y = systematic_sample.Class

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=1)

ml_model(X_train, X_test, Y_train, Y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Model Name  Train Accuracy  Test Accuracy  Precision  Recall  \
4  RandomForestClassifier            1.00           1.00       1.00    1.00   
2  DecisionTreeClassifier            1.00           0.99       0.99    0.99   
0      LogisticRegression            0.97           0.97       0.94    1.00   
3              GaussianNB            0.93           0.90       0.82    0.99   
1    KNeighborsClassifier            0.90           0.85       0.78    0.93   
5                     SVC            0.74           0.72       0.71    0.64   

   F1 score  
4      1.00  
2      0.99  
0      0.97  
3      0.90  
1      0.85  
5      0.67

## Stratified Sampling

In [59]:
statified_sample = df2.groupby('Class', group_keys=False).apply(lambda x: x.sample(frac=0.6))

In [60]:
X = statified_sample.loc[ : , statified_sample.columns != 'Class']
Y = statified_sample.Class

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=1)

ml_model(X_train, X_test, Y_train, Y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Model Name  Train Accuracy  Test Accuracy  Precision  Recall  \
4  RandomForestClassifier            1.00           0.99       0.99    0.99   
2  DecisionTreeClassifier            1.00           0.98       0.96    0.99   
0      LogisticRegression            0.94           0.94       0.88    0.97   
1    KNeighborsClassifier            0.90           0.86       0.76    0.95   
3              GaussianNB            0.78           0.75       0.62    0.95   
5                     SVC            0.73           0.71       0.61    0.73   

   F1 score  
4      0.99  
2      0.97  
0      0.92  
1      0.84  
3      0.75  
5      0.67

## Cluster Sampling

In [61]:
def cluster_sampling(df, number_of_clusters):

    try:
        # Divide the units into cluster of equal size
        df['cluster_id'] = np.repeat(
            [range(1, number_of_clusters+1)], len(df)/number_of_clusters)

        # Create an empty list
        indexes = []

        # Append the indexes from the clusters that meet the criteria
        # For this formula, clusters id must be an even number
        for i in range(0, len(df)):
            if df['cluster_id'].iloc[i] % 3 == 0:
                indexes.append(i)
        cluster_sample = df.iloc[indexes]
        return (cluster_sample)

    except:
        print("The population cannot be divided into clusters of equal size!")

In [62]:
cluster_sample = cluster_sampling(df2, 11)

In [63]:
X = cluster_sample.loc[ : , cluster_sample.columns != 'Class']
Y = cluster_sample.Class

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=1)

ml_model(X_train, X_test, Y_train, Y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Model Name  Train Accuracy  Test Accuracy  Precision  Recall  \
0      LogisticRegression            1.00           0.98       0.99    0.99   
3              GaussianNB            0.98           0.98       0.99    0.99   
4  RandomForestClassifier            1.00           0.98       1.00    0.97   
2  DecisionTreeClassifier            1.00           0.95       0.96    0.97   
1    KNeighborsClassifier            0.88           0.85       0.83    0.99   
5                     SVC            0.73           0.75       0.73    1.00   

   F1 score  
0      0.99  
3      0.99  
4      0.99  
2      0.97  
1      0.90  
5      0.84

## Result


| MODEL NAME             | Random Sampling | Systematic Sampling | Stratified Sampling | Cluster Sampling |
|------------------------|-----------------|---------------------|---------------------|------------------|
| RandomForestClassifier | 1.00            | 1.00                | 0.99                | 0.98             |
| DecisionTreeClassifier | 0.97            | 0.99                | 0.98                | 0.97             |
| LogisticRegression     | 0.94            | 0.97                | 0.94                | 0.97             |
| GaussianNB             | 0.92            | 0.91                | 0.91                | 0.94             |
| SVC                    | 0.77            | 0.84                | 0.82                | 0.86             |
| KNeighborsClassifier   | 0.74            | 0.74                | 0.74                | 0.75             |



As we can clearly see, ***RandomForestClassifier*** is the best model among the models we comapred